File: 1005-maximize-sum-of-array-after-k-negations/1005-maximize-sum-of-array-after-k-negations.py
class Solution:
    def largestSumAfterKNegations(self, nums: List[int], k: int) -> int:
        nums.sort()

        i = 0
        while i < len(nums) and k > 0 and nums[i] < 0:
            nums[i] = -nums[i]
            k -= 1
            i += 1
        
        minVal = float('inf')
        for num in nums:
            minVal = min(minVal, num)
        
        return sum(nums) - (k % 2) * minVal * 2

        

File: 1005-maximize-sum-of-array-after-k-negations/1005-maximize-sum-of-array-after-k-negations.java
class Solution {
    public int largestSumAfterKNegations(int[] nums, int k) {
        nums = IntStream.of(nums).boxed().sorted((a, b) -> Math.abs(b) - Math.abs(a))
        .mapToInt(Integer::intValue).toArray();
        
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] < 0 && k > 0) {
                nums[i] = - nums[i];
                k--;
            }
        }

        if (k % 2 == 1) {
            nums[nums.length - 1] = - nums[nums.length - 1];
        }
        return Arrays.stream(nums).sum();
    }
}